### PR TITLE
Cleaning up the user-facing API of FNAS.

### PR DESCRIPTION
**Pull request by dkorolev**
Hi @mzhurovich,

I started looking into optimizing member functions along with the static ones, and uniting various optimization algorithms under one cover, but ended up cleaning up the `typedef`-s for differentiable types first.

Please take a look.

Thanks,
Dima

CC @sompylasar @deathbaba.

### CODE REVIEW

**Comment by dkorolev** [comment]
Good news everyone! I got rid of `reset_internals_singleton()`.

The user has to keep two things in mind (`TODO(dkorolev)`: Document them on the Wki).
1. Only one expression per thread can be evaluated at a time.
2. The instance of `fncas::X` should be in the scope throughout the whole period of using the `FNCAS`-ed function.

@mzhurovich : PTAL. Thanks!

File: fncas/node.h [hunk 1]
```diff
@@ -356,33 +377,52 @@ struct f_intermediate : f {
...
+using fncas::V2X;
+
 // Arithmetic operations and mathematical functions are defined outside namespace fncas.
 
 #define DECLARE_OP(OP, OP2, NAME)                                                    \
-  inline fncas::node operator OP(const fncas::node& lhs, const fncas::node& rhs) {   \
-    fncas::node result;                                                              \
+  inline fncas::V operator OP(const fncas::V& lhs, const fncas::V& rhs) {   \
+    fncas::V result;                                                              \
     result.type() = fncas::type_t::operation;                                        \
```
Comment (mzhurovich) [inline fncas/node.h]: Formatting :(

File: fncas/node.h [hunk 2]
```diff
@@ -394,14 +434,15 @@ DECLARE_OP(*, *=, multiply);
 DECLARE_OP(/, /=, divide);
 
 #define DECLARE_FUNCTION(F)                           \
-  inline fncas::node F(const fncas::node& argument) { \
-    fncas::node result;                               \
+  inline fncas::V F(const fncas::V& argument) { \
+    fncas::V result;                               \
     result.type() = fncas::type_t::function;          \
```
Comment (mzhurovich) [inline fncas/node.h]: Same here

File: deprecated_test/function.h [hunk 3]
```diff
@@ -57,7 +57,7 @@ void register_function(const char* name, T* impl) {
 #define REGISTER_FUNCTION(F)                                                                              \
   struct enhanced_##F : F {                                                                               \
     virtual double eval_as_double(const std::vector<double>& x) const { return F::f(x); }                 \
-    virtual fncas::output<fncas::x>::type eval_as_expression(const fncas::x& x) const { return F::f(x); } \
+    virtual fncas::V eval_as_expression(const fncas::X& x) const { return F::f(x); } \
```
Comment (mzhurovich) [inline deprecated_test/function.h]: Looks like `clang_format` doesn't like your macros :(

**Comment by dkorolev** [comment]
Actually, I did not run `clang-format`, since `make indent` didn't work, since the `Makefile` is not a symlink, and our broken `readlink`-related script failed. Pichalka.

@mzhurovich : PTAL.

**Comment by dkorolev** [comment]
@mzhurovich: Could you take a look please? I'm thinking of merging it in and asking for a post-review to simplify deploying the presentation.

Thanks!
Dima

**Comment by dkorolev** [comment]
I'll merge this change for post-review. CC @mzhurovich.

Sorry for doing this. There's a mess with forked repository names (KnowSheet/fncas vs. old dkorolev/fncas vs. dkorolev/fncas which is a fork from KnowSheet), and I'd like to resolve it once and for all while deploying the demo for today.

Thanks!
Dima

**Comment by mzhurovich** [comment]
@dkorolev Ok, I'll look at the code today and let you know if I find something worth to mention

**Pull request merged**